tests/test_tls.c

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/http/connection.h>
+#include <aws/http/request_response.h>
+
+#include <aws/common/clock.h>
+#include <aws/common/condition_variable.h>
+#include <aws/common/mutex.h>
+#include <aws/io/channel_bootstrap.h>
+#include <aws/io/event_loop.h>
+#include <aws/io/host_resolver.h>
+#include <aws/io/logging.h>
+#include <aws/io/socket.h>
+#include <aws/io/tls_channel_handler.h>
+#include <aws/io/uri.h>
+#include <aws/testing/aws_test_harness.h>
+
+/* Singleton used by tests in this file */
+struct test_ctx {
+    struct aws_allocator *alloc;
+    struct aws_logger logger;
+    struct aws_event_loop_group event_loop_group;
+    struct aws_host_resolver host_resolver;
+    struct aws_tls_ctx *tls_ctx;
+    struct aws_client_bootstrap *client_bootstrap;
+    struct aws_http_connection *client_connection;
+    struct aws_http_stream *stream;
+
+    size_t body_size;
+    bool stream_complete;
+    bool client_connection_is_shutdown;
+
+    struct aws_mutex wait_lock;
+    struct aws_condition_variable wait_cvar;
+    int wait_result;
+};
+
+static const size_t TEST_TIMEOUT_SEC = 20;
+
+void s_on_connection_setup(struct aws_http_connection *connection, int error_code, void *user_data) {
```

Comment:
Something seems off about the concurrency here.  Main thread does a lock - cond wait - unlock as expected, but these notification callbacks never do a lock before modifying shared state and then they do an unlock which is maybe undefined since there was no corresponding lock?  Also mutex wasn't inited/cleaned up in the timeout test either, although bitwise zero is essentially init for our current platforms.

Reply:
Good catch, the mutex was an omission of copy-pasta, and the mutex/cvar init/clean_up was missing from the first test...yow